### PR TITLE
JW-1123 Time logger for build time tracking.

### PR DESCRIPTION
**Pull request by m-kostrzewa**

### CODE REVIEW

**Comment by m-kostrzewa** [comment]
You can see script output @ the bottom of: http://ci.wincontrail.codilime.com:8080/job/Wincontrail/job/build-all/351/console if you ctrl+f "Time measurement results"

File: CIScripts/BuildFunctions.ps1 [hunk 1]
```diff
@@ -69,61 +73,73 @@ function Invoke-DockerDriverBuild {
...
-    go build -v $DriverSrcPath
+    $TimeLogger.LogAndMeasure("Installing test runner", {
+        go get -u -v github.com/onsi/ginkgo/ginkgo
+    })
 
-    $srcPath = "$Env:GOPATH/src/$DriverSrcPath"
-    Write-Host $srcPath
+    $TimeLogger.LogAndMeasure("Building driver", {
+        go build -v $DriverSrcPath
+        Write-Host $srcPath
```
Comment (sodar) [inline CIScripts/BuildFunctions.ps1]: Is this `Write-Host` necessary here? If so, could you move it to the place where `$srcPath` is assigned.

File: CIScripts/TimeLogger.ps1 [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        $tm.Start = Get-Date
+        $this.Stack.Push($tm)
+        $this.CurrentIndentLevel += $this.IndentJump
+    }
+
+    Pop() {
+        $tm = $this.Stack.Pop()
+        $tm.End = Get-Date
+        $this.CurrentIndentLevel -= $this.IndentJump
+        $this.Output += ($tm.GetResultString($this.CurrentIndentLevel))
```
Comment (sodar) [inline CIScripts/TimeLogger.ps1]: Would it be possible to reverse output printing? Currently it is presented like this:

```
<other results>
        - 00:03:19.5284725: Copying Agent dependencies
        - 00:00:20.2641599: Building API
        - 00:23:36.7044592: Building contrail-vrouter-agent.exe and .msi
        - 00:00:00.0310005: Building tests
- 00:27:16.5360921: Agent build
<other results>
```

But it makes more sense to print it like this:

```
<other results>
- 00:27:16.5360921: Agent build
        - 00:03:19.5284725: Copying Agent dependencies
        - 00:00:20.2641599: Building API
        - 00:23:36.7044592: Building contrail-vrouter-agent.exe and .msi
        - 00:00:00.0310005: Building tests
<other results>
```

**Review by sodar — DISMISSED** [review]

**Review by jacek-izykowski — DISMISSED** [review]

**Comment by m-kostrzewa** [comment]
Janie, przetestuj proszę

**Comment by m-kostrzewa** [comment]
Janie, przetestuj proszę

**Review by sodar — DISMISSED** [review]

File: CIScripts/BuildFunctions.ps1 [hunk 3]
```diff
@@ -138,34 +156,52 @@ function Invoke-ExtensionBuild {
...
+        # TODO: Add other tests here once they are functional.
+        $Tests = @("agent:test_ksync", "agent:test_vnswif", "src/ksync:ksync_test")
+
+        $TestsString = ""
+        if ($Tests.count -gt 0) {
+            $TestsString = $Tests -join " "
+        }
+        $BuildCommand = "scons contrail-vrouter-agent.msi -j 4"
+        $AgentAndTestsBuildCommand = "{0} {1}" -f "$BuildCommand", "$TestsString"
+        Invoke-Expression $AgentAndTestsBuildCommand
```
Comment (jacek-izykowski) [inline CIScripts/BuildFunctions.ps1:190]: Target `contrail-vrouter-agent.msi` is invoked second time here.

File: CIScripts/BuildFunctions.ps1 [hunk 4]
```diff
@@ -138,34 +156,52 @@ function Invoke-ExtensionBuild {
...
-    }
+    $TimeLogger.LogAndMeasure("Building contrail-vrouter-agent.exe and .msi", {
+        scons contrail-vrouter-agent.msi -j 4
+        if ($LASTEXITCODE -ne 0) {
+            throw "Building Agent failed"
+        }
+    })
+
+    $TimeLogger.LogAndMeasure("Building tests", {
+        $Tests = @()
```
Comment (jacek-izykowski) [inline CIScripts/BuildFunctions.ps1]: It doesn't seem to serve any purpose.

File: CIScripts/TimeLogger.ps1 [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    TimeLogger () {
+        $this.Root = [TimeMeasurement]::new("Total")
+        $this.Root.Start = Get-Date
+        $this.Stack = New-Object System.Collections.Stack
+        $this.Stack.Push($this.Root)
+        $this.CurrentIndentLevel = 0
+    }
+
+    LogAndPush([string] $msg) {
```
Comment (jacek-izykowski) [inline CIScripts/TimeLogger.ps1]: Do we want to follow MS' encouraged development guidelines for PowerShell (coding convention)?
[https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx#SD03](url)

There are a few places that require attention in regards to those guidelines (if we decide to follow them).

**Review by jacek-izykowski — DISMISSED** [review]

File: CIScripts/TimeLogger.ps1 (hunk 5 above)
Comment (m-kostrzewa) [reply CIScripts/TimeLogger.ps1]: These guidelines are for cmdlets (like `Get-Service`, `Get-VMSwitch`), not for classes and methods, unless I'm missing something?

**Comment by m-kostrzewa** [comment]
Janie, przetestuj proszę

**Comment by m-kostrzewa** [comment]
Janie, przetestuj proszę

**Comment by m-kostrzewa** [comment]
Janie, przetestuj proszę

**Review by sodar — APPROVED** [review]
Looks good to me

**Comment by m-kostrzewa** [comment]
Janie, przetestuj proszę

**Pull request merged**